[SYCL] Make sycl::ext::oneapi::atomic_ref available in sycl namespace

psamolysov-intel opened the pull request:
The SYCL 2020 specification defines the sycl::atomic_ref template class
(section 4.15.3). But the DPC++ compiler defines the template class
under the sycl::ext::oneapi namespace.

Even though generic_space is added into the access::address_space
enumeration and used as the default value for the AddressSpace template
parameter of the sycl::atomic_ref class, the generic address space is not
supported yet. A compilation error occurs when the sycl::atomic_ref
template class is instantiated with generic_space as the address space.

The sycl::ext::oneapi::atomic_ref is marked as deprecated. An integration
test were added to check the deprecation warning.

psamolysov-intel commented:
The PR https://github.com/intel/llvm-test-suite/pull/545 must be merged before to let `llvm-test-suite` be tested against the new version of the compiler.

Pennycook commented on sycl/include/CL/sycl/atomic.hpp:
```diff
@@ -21,8 +21,9 @@
 #define __SYCL_STATIC_ASSERT_NOT_FLOAT(T)                                      \
-  static_assert(!std::is_same<T, float>::value,                                \
-                "SYCL atomic function not available for float type")
+  static_assert(                                                               \
+      !std::is_same<T, float>::value && !std::is_same<T, double>::value,       \
+      "SYCL atomic function not available for float and double types")
```
We shouldn't add `double` support to the legacy `atomic` types.  These were deprecated in SYCL 2020 (https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:atom-types-depr), and we shouldn't add new functionality to deprecated features.

psamolysov-intel replied:
@Pennycook Thank you for the answer. I've reverted the commit that adds `double` to the `atomic.hpp` header.

Pennycook commented on sycl/include/CL/sycl/atomic_ref.hpp:
```diff
@@ -0,0 +1,668 @@
+#endif
+
+template <typename T>
+using IsValidAtomicRefType =
+    bool_constant<IsValidAtomicType<T>::value || std::is_pointer<T>::value>;
```
This will have to change when `double` drops out of `IsValidAtomicType`.

psamolysov-intel replied:
@Pennycook Yes, I've changed both definitions of `IsValidAtomicRefType`, in the `sycl::` as well as in `sycl::ext::oneapi::` namespaces.

Pennycook commented on sycl/include/CL/sycl/atomic_ref.hpp:
```diff
@@ -0,0 +1,668 @@
+};
+
+// Functionality for any atomic of type T, reused by partial specializations
+template <typename T, memory_order DefaultOrder, memory_scope DefaultScope,
+          access::address_space AddressSpace>
```
The SYCL 2020 version of `atomic_ref` defaults this last template argument (see https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:atomic-references):
```c++
template <typename T, memory_order DefaultOrder, memory_scope DefaultScope,
          access::address_space Space = access::address_space::generic_space>
class atomic_ref;
```

I'm not sure if `access::address_space::generic_space` is defined yet, so you may need to add it.

We should probably add some new tests using atomics with the generic address space, as well.

psamolysov-intel replied:
@Pennycook `access::address_space::generic_space` is not defined yet, I've define it but without any logic, just a variant in the enumeration. Also, `sycl::atomic_ref` was modified to use `access::address_space::generic_space` as the default value for `AddressSpace` template parameter. Please, have a look.

Pennycook reviewed (approved):
Changes here look good to me, but I still think we need tests of the new `sycl::atomic_ref` (especially when using generic address space).

dm-vodopyanov reviewed (commented):
Please add tests for sycl::atomic_ref and fix already existed sycl::ext::oneapi::atomic_ref tests.

psamolysov-intel commented:
@dm-vodopyanov I've fixed the existing test here: https://github.com/intel/llvm-test-suite/pull/545 After the fix has been merged, the precommit checks should pass and I'll add the tests for `sycl::atomic_ref` in `llvm-test-suite`.

dm-vodopyanov commented:
@psamolysov-intel Please add a link to PR with new tests to this thread.

psamolysov-intel commented:
@dm-vodopyanov I'm working on the tests in this PR: https://github.com/intel/llvm-test-suite/pull/548 Exactly, the checks will fail since there is no `sycl::atomic_ref` class in the `intel` branch of the compiler yet.

dm-vodopyanov commented:
> @dm-vodopyanov I'm working on the tests in this PR: intel/llvm-test-suite#548 Exactly, the checks will fail since there is no sycl::atomic_ref class in the intel branch of the compiler yet.

Thanks, that's expected.

steffenlarsen reviewed (dismissed):
Minor note, but otherwise LGTM!

psamolysov-intel commented:
@Pennycook I have a question about the existence of two classes: `sycl::atomic_ref` and `sycl::ext::oneapi::atomic_ref`. I've removed the `IsValidAtomicType` alias and instead introduce the `IsValidAtomicRefType` one for the `sycl::ext::oneapi::atomic_ref` extension and now the extension won't be compiled with `wchar_t` for example. Also, the code of `sycl::ext::oneapi::atomic_ref` is just a duplication of the code of `sycl::atomic_ref` except the `AddressSpace` template parameter for `sycl::atomic_ref` has a default value.

Should I remove the `IsValidAtomicType` alias for `sycl::ext::oneapi::atomic_ref` to save the behavior of the extension and do not break any user code? If not, maybe the `sycl::ext::oneapi::atomic_ref` class should be just an alias for `sycl::atomic_ref`? As I see, usually we define old types as aliases for the new ones when we move it from a namespace to another namespace or from a header to another header.

psamolysov-intel commented on sycl/include/CL/sycl/atomic_ref.hpp:
```diff
@@ -0,0 +1,675 @@
+using memory_order = cl::sycl::memory_order;
+using memory_scope = cl::sycl::memory_scope;
+
+#if __cplusplus < 201703L
```
This code is required for C++ < 17 (C++14 in our case) since `memory_enums.hpp` defines such constants for C++17 only. This is not a protection from `inline constexpr`.

steffenlarsen replied:
Could there be a benefit to removing `if __cplusplus >= 201703L` in both this and memory_enums.hpp, then replace `inline constexpr` in memory_enums.hpp with `__SYCL_INLINE_CONSTEXPR`?

Pennycook replied:
It's for consistency with ISO C++.  C++11 had an unscoped enum, and C++20 swapped it for a scoped enum.  My understanding is that the `inline constexpr` versions were introduced for backwards compatibility with the old names. 
 See https://en.cppreference.com/w/cpp/atomic/memory_order.

We could have said that SYCL 2020 had an unscoped enum (for consistency with C++17) and then had SYCL-Next define `inline constexpr` versions for backward compatibility (for consistency with C++20), but since `inline constexpr` works in C++17 there didn't seem to be any harm in aligning with the C++20 version right away.

psamolysov-intel replied:
@steffenlarsen @gmlueck @Pennycook As I understand, it's better to keep things as they are today?

psamolysov-intel replied:
@gmlueck This code is protected from C++17, not from C++14 (**<** 201703L), but the code is required only because the constants is used inside the implementation of `sycl::atomic_ref`. I think it is better to remove this code and use `memory_order::relaxed` instead of `memory_order_relaxed`, etc in the class.

gmlueck replied:
Oh, I read the code too quickly.  I'm confused about the macro test now.  Why isn't the test `#if __cplusplus >= 201703L`?

Yes, it makes sense to use `memory_order::relaxed` instead of `memory_order_relaxed` in the implementation of the header.  The SYCL spec still requires the `memory_order_relaxed` variables to be defined, though.

psamolysov-intel replied:
@gmlueck This code added the constants for C++14 only since for C++17 (and newer) they are added by `sycl/memory_enums.hpp`. I've removed this code. Also, could you have a look at my latest question in the PR (about `generic_space`)?

psamolysov-intel commented:
@Pennycook I agree, the fact that we have specified a default value for a template parameter in `sycl::atomic_ref` should not touch the users of `sycl::ext::oneapi::atomic_ref` at all: they passed some value for that parameter since there were no default value for it.

psamolysov-intel commented:
I'm going to define `sycl::ext::oneapi::atomic_ref` as an alias for `sycl::atomic_ref` and solve problems with the generic address space.

psamolysov-intel commented:
@Pennycook @gmlueck I've found the following problem with the generic address space. Clang treats the `__attribute__((opencl_generic))` as nothing, therefore the following code:

```c++
#define __OPENCL_GENERIC_AS__ __attribute__((opencl_generic))
...
template <typename ElementType>
struct DecoratedType<ElementType, access::address_space::generic_space> {
  using type = __OPENCL_GENERIC_AS__ ElementType;
};
```
is just equals to the following:

```c++
struct DecoratedType<ElementType, access::address_space::generic_space> {
  using type = ElementType;
};
```
and therefore there are two equivalent constructors in the `multi_ptr.hpp`:

```c++
using pointer_t = typename detail::DecoratedType<ElementType, Space>::type *;
...
#ifdef __SYCL_DEVICE_ONLY__
  multi_ptr(pointer_t pointer) : m_Pointer(pointer) {}
#endif
```
and

```c++
  multi_ptr(ElementType *pointer) : m_Pointer((pointer_t)(pointer)) {
    // TODO An implementation should reject an argument if the deduced
    // address space is not compatible with Space.
  }
```
since `typename detail::DecoratedType<ElementType, Space>::type *` is just ` ElementType *`. And I get an error about constructor duplication, the same is true for the assignment operators.

In clang's `include/clang/Basic/AttrDocs.td` I see the following:
```
def OpenCLAddressSpaceGenericDocs : Documentation {
  let Category = DocOpenCLAddressSpaces;
  let Heading = "__generic, generic, [[clang::opencl_generic]]";
  let Content = [{
The generic address space attribute is only available with OpenCL v2.0 and later.
```

I have no idea how to enable OpenCL v2.0 or later, I've tried `-cl-std=CL3.0` as well as `-cl-std=CL2.0` but as I get it  makes CLang to parse the C++ file as an OpenCL one, so no templates and other C++ features work.

In clang's Sema I've found the following code, I'm not sure is it related to the problem but it looks like we just ignore the `opencl_generic` attribute:

```c++
  /// If this is an OpenCL address space attribute, returns its SYCL
  /// representation in LangAS, otherwise returns default address space.
  LangAS asSYCLLangAS() const {
    switch (getKind()) {
    ...
    case ParsedAttr::AT_OpenCLGenericAddressSpace:
    default:
      return LangAS::Default;
    }
  }
```

Also I read the SYCL 2020 specification [4.7.7.1 Multi-pointer class](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:multiptr) and see we have no `access::decorated` enumerator; therefore, we cannot support the overall set of requirements for the `generic_space`, for example `sycl::multi_ptr` should have the third template parameter of the `access::decorated` type, and many explicit cast operators from the generic address space into the local, global, and private ones must be there as well. But I cannot compile a minimal piece of code that instantiates a `sycl::atomic_ref` variable with the generic address space since when the `__attribute__((opencl_generic))` attribute is in use on the device side, clang sees no difference between the `ElementType *`  type and the `pointer_t` one. 

Is this behavior what is expected and a workaround must be provided, for example - to define [the constructor](https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl/multi_ptr.hpp#L57-L60) and [assignment operator](https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl/multi_ptr.hpp#L80-L85) for host only?

Pennycook commented:
> @Pennycook I agree, the fact that we have specified a default value for a template parameter in `sycl::atomic_ref` should not touch the users of `sycl::ext::oneapi::atomic_ref` at all: they already passed some value for that parameter since there were no default value for it.

Sorry, that's the opposite of what I said.  If you define an alias, won't that allow somebody to use `sycl::ext::oneapi::atomic_ref` without passing a value for that parameter, or passing a value for `generic_address`?  I'd like us to deprecate and remove the old class eventually, and I don't understand why we'd go through the intermediate step of introducing an alias.

@gmlueck, what do you think?  Am I worried about an unimportant corner case?

Pennycook commented:
Tagging @rolandschulz as an expert on `multi_ptr` in SYCL 2020.

> And I get an error about constructor duplication, the same is true for the assignment operators.

Can't you define a new partial specialization of `multi_ptr` for the generic address space, where the decorated pointer type is equivalent to a raw pointer and the functions are only defined once?

> Also I read the SYCL 2020 specification 4.7.7.1 Multi-pointer class and see we have no access::decorated enumerator; therefore, we cannot support the overall set of requirements for the generic_space, for example sycl::multi_ptr should have the third template parameter of the access::decorated type, and many explicit cast operators from the generic address space into the local, global, and private ones must be there as well. 

Are you saying that we can never implement this, or just that it's going to require bigger changes to `multi_ptr`?

If you introduce a new partial specialization for `sycl::atomic_ref` and the generic address space, can you avoid using `multi_ptr` in the implementation?  That might allow you to focus on the `atomic_ref` changes for now.

> Is this behavior what is expected and a workaround must be provided, for example - to define the constructor and assignment operator for host only (but this idea won't work)?

I think we should try to avoid these sorts of workarounds, and especially when they involve `__SYCL_DEVICE_ONLY__`.  It's difficult to keep track of them, and I don't want us to get stuck with one here if we can find a way to implement the actual fix.

psamolysov-intel commented:
@Pennycook "Are you saying that we can never implement this, or just that it's going to require bigger changes to multi_ptr?"
Yes, this requires more work than just to introduce the `sycl::atomic_ref` class. 

About the specification of `multi_ptr` for `generic_space`. As I understand, the specification should contain only one constructor and one assignment operator since it is expected that the `pointer_t` alias in the generic address space is the same as `ElementType *`.

Thank you for the idea to introduce a partial specialization of `atomic_ref` for generic address space without `multi_ptr`. Should I use the row pointer there? I think it is better to be focused on `atomic_ref` in this PR.

psamolysov-intel commented:
@Pennycook @gmlueck About an alias for `sycl::ext::oneapi::atomic_ref`. Ups, sorry, my bad. But I see to have an alias is easy for the tests since having two classes breaks some tests in `llvm-test-suite` at least. This is not a big deal to fix them but we can avoid this. 

For now, the last template parameter of `sycl::ext::oneapi::atomic_ref` has no default value and the users have to explicitly pass some value there. If we have `generic_space` as the default value, one will have ability to use `sycl::ext::oneapi::atomic_ref` with this value but what problem it can create? If we leave `sycl::ext::oneapi::atomic_ref` as a class (not an alias) it just makes code duplication.

Pennycook commented:
> For now, the last template parameter of `sycl::ext::oneapi::atomic_ref` has no default value and the users have to explicitly pass some value there. If we have `generic_space` as the default value, one will have ability to use `sycl::ext::oneapi::atomic_ref` with this value but what problem it can create? 

I'm worried that we'd be implicitly adding new features to a deprecated API.  I also thought that redefining `atomic_ref` as an alias to a class in a different namespace would be an ABI break, but I'm not sure about that.

> If we leave `sycl::ext::oneapi::atomic_ref` as a class (not an alias) it just makes code duplication.

If the intent is to eventually delete `sycl::ext::oneapi::atomic_ref` completely, the duplicated code would only exist as long as the deprecation message.  It seems to me that adding the alias is more work than just creating a copy.

AlexeySachkov commented on sycl/include/CL/sycl/atomic_ref.hpp:
```diff
@@ -0,0 +1,662 @@
+
+// DefaultOrder parameter is limited to read-modify-write orders
+template <memory_order Order>
+using IsValidDefaultOrder = bool_constant<Order == memory_order::relaxed ||
```
If I understand correctly, [`bool_constant`](https://en.cppreference.com/w/cpp/types/integral_constant) resides in `std` namespace. The fact that you are able to use it here without `std::` concerns me, because it assumes that there is a `using namespace std;` somewhere before this header is included.

UPD: I've realized that we have `bool_constant` in our `sycl::detail` namespace:
https://github.com/intel/llvm/blob/0c33048e492661e50596eb5f8650eeff92347db9/sycl/include/CL/sycl/detail/stl_type_traits.hpp#L39-L40

That is probably a minor thing, but it still confusing: it is not clear which version is intended to be used here, we don't have explicit include for an internal one, etc.

psamolysov-intel replied:
Since the `bool_constant` is used here, in the `sycl::detail` namespace, the definition from the same namespace will be used, so our crafted `bool_constant`. As I understand, we defined `sycl::detail::bool_constant` since this alias in the `std` namespace has been defined since C++17 (https://en.cppreference.com/w/cpp/types/integral_constant) but we also should support C++14 too.

AlexeySachkov commented on sycl/include/sycl/ext/oneapi/atomic_enums.hpp:
```diff
@@ -24,31 +24,40 @@ namespace sycl {
 #ifndef __SYCL_DEVICE_ONLY__
 namespace detail {
 
+__SYCL2020_DEPRECATED(
+    "use 'sycl::detail::getStdMemoryOrder(sycl::memory_order)' instead")
```
Wow, that is something new to me. In which cases such diagnostic is going to be visible? I mean when user compiles their SYCL application, our headers are marked as system ones and they shouldn't produce "internal" diagnostics, i.e. warnings for things which are not directly used in user code.

Will it be visible during DPC++ RT build? If so, do we have a `-Werror` build which will fail (@alexbatashev?) ?

psamolysov-intel replied:
As I see, the [[deprecated]] attribute for the alias introduced by the `using` keyword will be ignored by the compiler, since the compiler just substitutes the target of the alias. So this concrete deprecation is for documentation purposes only. But the [[deprecated]] attribute next to the `memory_order_relaxed` and the other constants generates a warning for the user, please have a look at the test https://github.com/psamolysov-intel/llvm/blob/make-sycl-atomic-ref-available/sycl/test/basic_tests/ext-atomic-ref-instantiation.cpp

alexbatashev replied:
> Will it be visible during DPC++ RT build?

No, we disable all deprecation warnings during runtime build

AlexeySachkov replied:
> No, we disable all deprecation warnings during runtime build

What is the point then of this deprecation? It won't help find usage of deprecated internal function? Or am I missing something?

dm-vodopyanov replied:
Probably it should be removed as user shouldn't use "private" API under sycl::detail:: namespace.

psamolysov-intel replied:
@dm-vodopyanov Let's keep it there since just a marker the API is deprecated. In the future, the entire header will be removed as deprecated.

AlexeySachkov commented on sycl/include/CL/sycl/atomic_ref.hpp:
```diff
@@ -0,0 +1,662 @@
+#include <CL/sycl/detail/defines.hpp>
+#include <CL/sycl/detail/spirv.hpp>
+#include <CL/sycl/memory_enums.hpp>
+#ifdef __SYCL_DEVICE_ONLY__
+#include <CL/sycl/multi_ptr.hpp>
+#endif
```
Is this some kind of an optimization? If so, should we move include of `CL/sycl/detail/spirv.hpp` into this condition as well?

psamolysov-intel replied:
@AlexeySachkov Thank you for the idea, I've moved the import under the `__SYCL_DEVICE_ONLY__` guard.

AlexeySachkov commented on sycl/include/CL/sycl/atomic_ref.hpp:
```diff
@@ -0,0 +1,662 @@
+
+#pragma once
+
+#include <CL/__spirv/spirv_ops.hpp>
```
Do we really need this include? It seems like we are using all SPIR-V functions through `detail/spirv.hpp` helper

psamolysov-intel replied:
The header is included by `multi_ptr.hpp` too, so I've removed the include directive.

AlexeySachkov reviewed (dismissed):
`spirv.hpp` changes LGTM.

psamolysov-intel commented:
@Pennycook Thank you for the opinion about aliasing `sycl::ext::oneapi::atomic_ref`. I think the ABI break is an argument. Also, if we just mark something as an alias using the `using` keyword and add the `[[deprecated]]` attribute for the alias, the compiler will ignore this attribute since it will substitute the target class instead of the alias, and will generate no warning for the customer. I've reverted the commit that converts the `sycl::ext::oneapi::atomic_ref` class into an alias.

psamolysov-intel commented:
@Pennycook @gmlueck @AlexeySachkov I've solved compilation problems with `sycl::multi_ptr` when `access::address_space::generic_space` is in use but in fact the `sycl::atomic_ref` is unusable with the `generic_space` address space since clang doesn't support the address space. For example, the `__spirv\spirv_ops.hpp` defines the `__SPIRV_ATOMICS` macro as:

```c++
#define __SPIRV_ATOMICS(macro, Arg)                                            \
  macro(__attribute__((opencl_global)), Arg)                                   \
      macro(__attribute__((opencl_local)), Arg)
```

and this generates following errors when `generic_space` is used:

```
\include\sycl\CL/sycl/detail/spirv.hpp:381:10: error: no matching
      function for call to '__spirv_AtomicIAdd'
  return __spirv_AtomicIAdd(Ptr, SPIRVScope, SPIRVOrder, Value);
         ^~~~~~~~~~~~~~~~~~
\include\sycl\CL/sycl/atomic_ref.hpp:295:27: note: in instantiation of
      function template specialization
      'sycl::detail::spirv::AtomicIAdd<int, sycl::access::address_space::generic_space>' requested here
    return detail::spirv::AtomicIAdd(ptr, scope, order, operand);
                          ^
basic_tests/atomic_ref_plus_test.cpp:33:24: note: in instantiation of member function
      'sycl::detail::atomic_ref_impl<int, sycl::memory_order::relaxed, sycl::memory_scope::device,
      sycl::access::address_space::generic_space>::fetch_add' requested here
        out[gid] = atm.fetch_add(Difference(1));

...

\include\sycl\CL/__spirv/spirv_ops.hpp:138:29: note: expanded from
      macro '__SPIRV_ATOMIC_IADD'
  extern SYCL_EXTERNAL Type __spirv_AtomicIAdd(                                \
                            ^
\include\sycl\CL/__spirv/spirv_ops.hpp:251:17: note: candidate function
      not viable: no known conversion from 'int *' to '__global unsigned long long *' for 1st argument
__SPIRV_ATOMICS(__SPIRV_ATOMIC_UNSIGNED, unsigned long long)
                ^
\include\sycl\CL/__spirv/spirv_ops.hpp:241:3: note: expanded from macro
      '__SPIRV_ATOMICS'
  macro(__attribute__((opencl_global)), Arg)
```

If the `macro(__attribute__(opencl_generic)), Arg)` macro will be added, a lot of `llvm_unreachable` errors will occur in clang on the device side since clang doesn't support the `opencl_generic` attribute for SYCL yet (see log https://github.com/intel/llvm/runs/4165346494?check_suite_focus=true).

I think we should add a warning in the `atomic_ref.hpp` header about the `generic_space` address space is not supported yet, but I have no idea how to make the warning conditionally regarding to the `AddressSpace` template argument.

Pennycook commented:
That's disappointing.  Thanks for continuing to look into this.

> If the `macro(__attribute__(opencl_generic)), Arg)` macro will be added, a lot of `llvm_unreachable` errors will occur in clang on the device side since clang doesn't support the `opencl_generic` attribute for SYCL yet (see log https://github.com/intel/llvm/runs/4165346494?check_suite_focus=true).

Out of curiosity, what happens if you define these functions as accepting a raw pointer (without an address space attribute)?  I wonder if those cases would be converted into generic pointers correctly during the LLVM to SPIR-V conversion... But I have no idea if this will work -- just an idea that might be interesting to try.


> I think we should add a warning in the `atomic_ref.hpp` header about the `generic_space` address space is not supported yet, but I have no idea how to make the warning conditionally regarding to the `AddressSpace` template argument. We can add an error using `static_assert` but I think a warning is enough.

I don't know of any way to generate a warning conditionally.  Since the generated code wouldn't work, I think generating an error if the user tries to instantiate `atomic_ref` with a `generic_space` template argument would be acceptable.

psamolysov-intel commented:
@Pennycook Thank you for the suggestion. I've tried to add `macro(, Arg)` to let SPIR-V functions accept raw pointers and it seems to work (at least on Windows) but I'm not sure we should commit this changes, so I reverted the experiment. Also, I reverted the changes in `sycl::multi_ptr` to support `generic_space` address space since an error to `atomic_ref` is added when one is trying to use this address space. 

About the error when one is creating an `atomic_ref` instance with the generic address space, since the generic address space is not supported yet we can inform the user with an error. I added a `static_assert` to emit it. Could you review, please?

psamolysov-intel commented:
/summary:run

Pennycook reviewed (approved):
The error for the generic address space looks good to me.  I think this signals that we _will_ support the generic address space, but stops people from trying to use it for now.

dm-vodopyanov reviewed (approved):
Overall LGTM

bader commented on sycl/include/CL/sycl/atomic_ref.hpp:
```diff
@@ -0,0 +1,676 @@
+#endif
+#include <type_traits>
+
+__SYCL_INLINE_NAMESPACE(cl) {
```
> The SYCL 2020 specification defines the sycl::atomic_ref template class

Why did you put SYCL 2020 template class to `cl` namespace?

vladimirlaz replied:
@psamolysov-intel could you please fix that?

psamolysov-intel replied:
@bader @vladimirlaz I've removed the `cl` namespace wrapper there: https://github.com/intel/llvm/pull/4956

psamolysov-intel replied:
@bader @vladimirlaz When I tried to address the issue and move the `atomic_ref` class from `cl::sycl::` into the `sycl::` namespace, the following errors occurs:

```
sycl\sycl\unittests\helpers/CommonRedefinitions.hpp(149): error C2872: 'sycl': ambiguous symbol
sycl\sycl\include\CL/sycl/atomic_ref.hpp(25): note: could be 'sycl'
sycl\sycl\unittests\helpers/PiMock.hpp(38): note: or       'cl::sycl'
sycl\sycl\unittests\helpers/CommonRedefinitions.hpp(149): error C2039: 'unittest': is not a member of 'sycl'
sycl\sycl\include\CL/sycl/atomic_ref.hpp(25): note: see declaration of 'sycl'
sycl\sycl\unittests\helpers/CommonRedefinitions.hpp(149): error C2872: 'sycl': ambiguous symbol
sycl\sycl\include\CL/sycl/atomic_ref.hpp(25): note: could be 'sycl'
...
```

Even though the `atomic_ref` itself is compiled fine, since the class is included into the `sycl/sycl.hpp` header, a lot of code becomes broken. Since we have introduced two namespaces: `cl::sycl::` and `sycl::` itself and `cl::` is an inline namespace, the ambiguity occurs: does `sycl` mean `cl::sycl` or just `sycl`? So, we must change all the `sycl` occurs to `cl::sycl` but as I understand our goal is to eventually replace all `cl::sycl` with `sycl` in the future. 

Let's declare `atomic_ref` in the `cl::sycl::` namespace as it is, since the `cl::` namespace is inline, the users can reference to `atomic_ref` by `sycl::atomic_ref` without any problems. Also we have no classes defined directly in `sycl::`, not `cl::sycl::` namespace yet, even extensions from `sycl/ext/intel|oneapi` are defined in the `cl::sycl::` namespace.

alexbatashev replied:
@psamolysov-intel FYI @romanovvlad and @denis-kabanov are working on transition to `sycl::` namespace.

gmlueck replied:
I thought we previously decided that all SYCL APIs are available in the `::cl::sycl` namespace when the application includes `<CL/sycl.hpp>`.  This is what the SYCL 2020 spec says:

> For compatibility with SYCL 1.2.1, SYCL provides another standard header file: <CL/sycl.hpp>, which can be included in place of <sycl/sycl.hpp>. In that case, all SYCL classes, constants, types and functions defined by this specification should exist within the `::cl::sycl` C++ namespace.

bader replied:
> I thought we previously decided that all SYCL APIs are available in the `::cl::sycl` namespace when the application includes `<CL/sycl.hpp>`. This is what the SYCL 2020 spec says:
> 
> > For compatibility with SYCL 1.2.1, SYCL provides another standard header file: <CL/sycl.hpp>, which can be included in place of <sycl/sycl.hpp>. In that case, all SYCL classes, constants, types and functions defined by this specification should exist within the `::cl::sycl` C++ namespace.

Does it apply to extensions like `atomic_ref` before they are promoted to SYCL core part?

gmlueck replied:
This conversation is about the version of `atomic_ref` that *is* promoted to core SYCL, though, right?

In any case, it would probably be confusing to users if extensions behaved differently than core SYCL APIs in this regard.  Unless it is difficult to implement, I'd think that extensions should be available in the `cl::sycl::ext` namespace when `<CL/sycl.hpp>` is included.

bader replied:
AFAIK, SYCL 2020 extensions are available only in `::sycl::` namespace. 
Some extensions added before SYCL 2020 are available in `cl::sycl::` namespace as well.

BTW, is there any plan to drop support for `cl::` namespace in the future?

gmlueck replied:
> Some extensions added before SYCL 2020 are available in cl::sycl:: namespace as well.

This seems confusing if it is not consistent.

> BTW, is there any plan to drop support for cl:: namespace in the future?

I think we should, but I'm not sure what the other Khronos members are thinking.
